ext/nio4r/monitor.c

```diff
@@ -117,6 +119,37 @@ static VALUE NIO_Monitor_initialize(VALUE self, VALUE io, VALUE interests, VALUE
     return Qnil;
 }
 
+static VALUE NIO_Monitor_setInterests(VALUE self, VALUE interests){
+    //Check Whether the Monitor is already Closed?
+    if(NIO_Monitor_is_closed(self) == Qfalse){
+        struct NIO_Monitor *monitor;
+
+        ID interests_id;
+
+        interests_id = SYM2ID(interests);
+        Data_Get_Struct(self, struct NIO_Monitor, monitor);
+
+        if(interests_id == rb_intern("r")) {
+            monitor->interests = EV_READ;
+        } else if(interests_id == rb_intern("w")) {
+            monitor->interests = EV_WRITE;
+        } else if(interests_id == rb_intern("rw")) {
+            monitor->interests = EV_READ | EV_WRITE;
+        } else {
+            rb_raise(rb_eArgError, "invalid event type %s (must be :r, :w, or :rw)",
+                RSTRING_PTR(rb_funcall(interests, rb_intern("inspect"), 0, 0)));
+        }
+        //Stop the watcher and restart it after set the variable.
+        ev_io_stop(monitor->selector->ev_loop, &monitor->ev_io);
+        rb_ivar_set(self, rb_intern("interests"), interests);
```

Comment:
I think you need to call `ev_io_set` after `ev_io_stop` to actually change the interests. Calling `rb_ivar_set` only changes the `@interests` instance variable of the Ruby monitor object, but the actual guts of the code are in libev.

Something like this, I would think:

```c
ev_io_set (monitor->ev_io, monitor->ev_io.fd, monitor->interests);
```

See:

http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#ANATOMY_OF_A_WATCHER